[Behaviortree.cpp] Add 4.9.0

ericriff opened the pull request:
### Summary
Add `behaviortree.cpp/4.9.0` and drop support for the older versions.

#### Motivation
Older versions had all its dependencies vendored in. I tried to unvendored all of them here, but it ended up being a mess https://github.com/conan-io/conan-center-index/pull/28464.

So I updated upstream to support opting out of the vendored dependencies https://github.com/BehaviorTree/BehaviorTree.CPP/pull/1012

These changes were made available on v4.8.0 and it greatly simplifies the packaging process.

The current version of the recipe is a complex combination of if/elses to unvendor some, but not all the dependencies. Making it compatible with this new version would add even more if/elses and make it very messy. Particularly because over time some dependencies changed (e.g. boost got replaced with minicoro).

So I've decided to start a new recipe from scratch. Alternatively I could keep the older versions on a separate subfolder, something like `pre_4.8.0`.

---
- [X] Read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md)
- [X] Checked that this PR is not a duplicate: [list of PRs by recipe](https://github.com/conan-io/conan-center-index/discussions/24240)
- [X] Tested locally with at least one configuration using a recent version of Conan

ericriff commented on recipes/behaviortree.cpp/all/conanfile.py:
```diff
@@ -169,12 +89,8 @@ def validate(self):
         if self.settings.compiler == "clang" and str(self.settings .compiler.libcxx) == "libstdc++":
             raise ConanInvalidConfiguration(f"{self.ref} needs recent libstdc++ with charconv. please switch to gcc, or to libc++")
 
-        if self.settings.compiler == "apple-clang" and cross_building(self) and self.settings.arch in ["x86", "x86_64"]:
```
It looks like this was outdated.. I removed it just to see what happens and CICD didn't complain so we should be good?

uilianries replied:
Totally fine!

uilianries commented on recipes/behaviortree.cpp/all/conanfile.py:
```diff
@@ -260,50 +148,33 @@ def package(self):
             bin_path = os.path.join(self.package_folder, "bin")
             self.env_info.PATH.append(bin_path)
 
         # TODO: to remove in conan v2 once cmake_find_package* generators removed
```
This entire block can be removed. It's a legacy related to the deprecated CMake generator from Conan 1.x

ericriff commented on recipes/behaviortree.cpp/all/conanfile.py:
```diff
@@ -169,12 +89,8 @@ def validate(self):
     def build_requirements(self):
-        if Version(self.version) >= "4.1.0":
-            self.tool_requires("cmake/[>=3.16.3 <4]")
+        self.tool_requires("cmake/[>=3.16.3 <4]")
```
I've no idea. I kept the restriction as it was. I'll do a local check before bumping it.

ericriff replied:
It does indeed work. I've bumped the max version as suggested here 731975d254b2ca80846c1ddde61d31c6611aae46.

uilianries commented:
@ericriff Hello! As this PR changed a lot in this recipe, and also the new release of Behavior.Tree CPP is now unvendored, please, try also to check those options like tools. 

Locally, I failed to build the project using tools: /home/uilian/Development/conan/conan-center-index/recipes/behaviortree.cpp/behaviortree.cpp-4.8.0-linux-gcc11-tools.log

There are several linkage errors regarding libsodium, a transitive dependency of zeromq. 

The main point is: this error only occurs when building tools. The tool `bt_recorder` explicitly uses ZeroMQ (not zmqpp): https://github.com/BehaviorTree/BehaviorTree.CPP/blob/4.8.0/tools/bt_recorder.cpp#L6

Both this PR and that in BehaviorTree changed ZeroMQ usage. Please, check what can be done to fix it.

ericriff commented:
I had to add a patch since this is a problem on the behaviorcppp side. As you stated that binary uses zeromq directly (instead of the cppzmq wrapper), but it doesn't even do a find_package for it.
I made zeromq a conditional requirement based on the value of with_tools and included a patch to handle it gracefully, hopefully on all situations.

ericriff commented:
So I did some more digging. Even though the patch I pushed here works, it is not the proper solution. The problem is that on my PR to un-vendor the libraries I inadvertently moved cppzmq and sqlite3 from PUBLIC to PRIVATE on the target_link_libraries call for btcpp. 
I opened a PR upstream but the maintainer did not take it. Instead he merged a different solution which does not fix our issues.
How should we proceed here? 
https://github.com/BehaviorTree/BehaviorTree.CPP/pull/1026

nyibbang commented:
Hello @ericriff
First, I'd like to thank you for your work on the recipe.
I'm about to start a new project with BehaviorTree.CPP & Conan but I would really love to pull the latest version of the library (especially because I saw they fixed a bunch of mem leaks and thread safety issues in 4.8.0).

Is there an update on this PR ? Is there something blocking, and can we help ?

ericriff commented:
Hey @nyibbang, lets ping @uilianries. Maybe this discussion fell through the cracks on his side.

Just to reiterate the problem: my original implementation on this PR passes all checks on CCI, but @uilianries identified an issue with some option combo that CCI doesn't validate.
I pushed a patch here (the last commit) that works locally, but later on found out it is not the proper fix. See my comment above.
I sent the proper fix upstream as a PR (https://github.com/BehaviorTree/BehaviorTree.CPP/pull/1026), but the maintainer of bt didn't take it. That surprised me a little bit because that PR fixes a bug I introduced on bt.cpp when I refactored their handling of 3rd party libs to make the conan recipe easier to deal with (coneofshame). But oh well.
So the options are:
1. Merge this PR with no patches and have a recipe that fails to build with some combination of options
2. Merge this PR with the current patch
3. Revert the current patch and instead use the one the maintainer of bt didn't take. I'd prefer that one.

uilianries commented on recipes/behaviortree.cpp/all/conanfile.py:
```diff
@@ -24,133 +23,60 @@ class BehaviorTreeCPPConan(ConanFile):
+        self.requires("minitrace/cci.20230905")
+        self.requires("tinyxml2/11.0.0")
+        if self.options.enable_groot_interface:
+            self.requires("cppzmq/4.11.0")
```
```suggestion
```

uilianries commented on recipes/behaviortree.cpp/config.yml:
```diff
@@ -1,17 +1,3 @@
-  "3.8.6":
-    folder: all
-  "3.7.0":
+  "4.8.0":
```
```suggestion
  "4.8.2":
```

uilianries reviewed (commented):
@ericriff Hello!

After doing a few new tests, including building BehaviorTree.CPP without this recipe, I think it would be better to keep it as close as possible to the upstream: Let's use its embedded cppmq instead of the Conan package, it will make the maintenance much simpler.

As a result, we no longer need to apply patches for this recipe. 

The release 4.8.2 is out too with a few build changes. Could you please bump your PR to follow the latest release? Regards! 

BTW, my current build log is here: [build.log](https://github.com/user-attachments/files/24165072/build.log)

ericriff commented:
@uilianries your PR upstream got accepted, nice!
I'd like to wrap up this PR. Would you be ok with me using your PR as a backport patch on this recipe? Or should we wait for a new release?

uilianries commented:
@ericriff Thank you for pinging me, I just missed that notification! 

Please, feel free to backport that PR here 😄

uilianries commented on recipes/behaviortree.cpp/all/conandata.yml:
```diff
@@ -1,34 +1,10 @@
-    - patch_file: "patches/4.6.2-001-minitrace.patch"
-      patch_description: "Consume minitrace from Conan"
-      patch_type: "conan"
+  "4.8.0":
```
```suggestion
  "4.8.2":
```

Version mismatch

uilianries commented on recipes/behaviortree.cpp/all/conanfile.py:
```suggestion
required_conan_version = ">=2.1"
```

Due `implements` feature

uilianries commented on recipes/behaviortree.cpp/all/conanfile.py:
```diff
@@ -24,133 +23,60 @@ class BehaviorTreeCPPConan(ConanFile):
-                "msvc": "191",
-                "Visual Studio": "15",
-            }
+        return {
```
This can be removed. In Conan 2.x cppstd is always available.

uilianries commented on recipes/behaviortree.cpp/all/conanfile.py:
```diff
@@ -184,65 +83,36 @@ def generate(self):
+        cmd.generate()
 
     def _patch_sources(self):
         apply_conandata_patches(self)
```
```suggestion
```

ericriff commented:
@uilianries could I get an approval here? I want to merge this before there is another release out haha

uilianries commented on recipes/behaviortree.cpp/all/conanfile.py:
```diff
@@ -260,50 +126,13 @@ def package(self):
-            requires.append("tinyxml2::tinyxml2")
-        if self._with_zeromq:
-            requires.append("zeromq::zeromq")
+        self.cpp_info.set_property("cmake_file_name", "BehaviorTree")
```
```suggestion
        self.cpp_info.set_property("cmake_file_name", "behaviortree_cpp")
```

- https://c3i.jfrog.io/artifactory/cci-build-logs/cci/prod/PR-28609/18/package_build_logs/build_log_behaviortree_cpp_4_8_3_0f96f60fddf22ee68f4503c6bc48cbd0_035f13d59d990f1e04c66590ae708dcc0452e373.success.txt
- https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/cmake/Config.cmake.in#L3

uilianries reviewed (approved):
LGTM

@ericriff Thank you for this huge effort, mainly for working on the PR https://github.com/BehaviorTree/BehaviorTree.CPP/pull/1012, which improved Behavior.Tree dependency management! 🚀 

----

For reviewers:

- Desvendorizantion was achieved on Behavior.Tree [4.8.0](https://github.com/BehaviorTree/BehaviorTree.CPP/releases/tag/4.8.0)
- Later patches versions fixed or improved some details, but the result can consume external dependencies. 
- The options `with_coroutines` and `enable_manual_selector` are no longer supported by the project since 4.1.0
- The [Config.cmake.in](https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/cmake/Config.cmake.in) shows the expected CMake file name and CMake target.

ericriff commented:
Do I need another approval here?

uilianries commented:
I just bumped to version 4.9.0: https://github.com/BehaviorTree/BehaviorTree.CPP/releases/tag/4.9.0

This new version dropped foonathan-lexy as a dependency: https://github.com/BehaviorTree/BehaviorTree.CPP/compare/4.8.4...4.9.0
